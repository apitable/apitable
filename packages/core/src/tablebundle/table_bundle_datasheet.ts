import { TableBundleSnapshot } from './table_bundle_snapshot';

export class TableBundleDataSheet{
    public snapshot: TableBundleSnapshot;
    public extras: any;

    public constructor(
      snapshot: TableBundleSnapshot,
      extras?: string
    ) {
      this.extras = extras;
      this.snapshot = snapshot;
    }
}