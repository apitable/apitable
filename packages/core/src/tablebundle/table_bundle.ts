import * as fflate from 'fflate';
import { TableBundleManifest } from './table_bundle_manifest';
import { TableBundleDataSheet } from './table_bundle_datasheet';
import { TableBundleSnapshot } from './table_bundle_snapshot';

export class TableBundle {
  public manifest: TableBundleManifest;
  private _nodeMap: Map<string, TableBundleDataSheet>;
  private _saver!: ITableBundleSaver;
  private _loader!: ITableBundleLoader;
  private constructor() {
    this.manifest = new TableBundleManifest();
    this._nodeMap = new Map<string, TableBundleDataSheet>();
  }

  public static new(options: ITableBundleInitOptions): TableBundle {
    const { loader, saver } = options;
    const tableBundle = new TableBundle();
    tableBundle.setTableBundleSaver(saver);
    tableBundle.setTableBundleLoader(loader);
    return tableBundle;
  }

  setTableBundleSaver(provider: ITableBundleSaver): void {
    this._saver = provider;
  }

  setTableBundleLoader(loader: ITableBundleLoader): void {
    this._loader = loader;
  }

  public nodeMap(): Map<string, TableBundleDataSheet> {
    return this._nodeMap;
  }

  public apply(sheet: TableBundleDataSheet, id: string = 'dst001', name: string = 'datasheet'): void {
    this._nodeMap.set(id, sheet);
    this.manifest.addDataSheetNode(id, name);
  }

  public getDataSheet(id: string): TableBundleDataSheet {
    return <TableBundleDataSheet>this._nodeMap.get(id);
  }

  save(path: string) {
    if (this._saver === undefined) {
      return;
    }
    const sheetData = {};
    this.nodeMap().forEach((value: TableBundleDataSheet, key: string) => {
      sheetData[key + '.json'] = fflate.strToU8(JSON.stringify(value.snapshot));
      if (value.extras !== undefined) {
        sheetData[key + '.extras.json'] = fflate.strToU8(value.extras);
      }
    });
    const zipped = fflate.zipSync({
      data: sheetData,
      'manifest.json': fflate.strToU8(JSON.stringify(this.manifest))
    }, {
      level: 1,
      mtime: new Date()
    });
    this._saver.save(zipped, path);
  }

  loadFile(path: string) {
    if (this._loader === undefined) {
      return;
    }
    this._nodeMap = new Map<string, TableBundleDataSheet>();
    const buffer = this._loader.load(path);
    const unzipped = fflate.unzipSync(buffer);
    if (unzipped['manifest.json'] === undefined) {
      throw new Error('manifest.json not found');
    }

    this.manifest = TableBundleManifest.build(JSON.parse(fflate.strFromU8(unzipped['manifest.json'])));
    const allChildNodes = this.manifest.allChildNodes();
    for (let i = 0; i < allChildNodes.length; i++) {
      const nodeId = allChildNodes[i];
      if (nodeId === undefined) {
        throw new Error('nodeId ' + nodeId + ' is undefined');
      }
      const fileName = 'data/' + nodeId + '.json';
      if (unzipped[fileName] === undefined) {
        throw new Error(fileName + ' not found');
      }
      // @ts-ignore
      const sn = JSON.parse(fflate.strFromU8(unzipped[fileName])) as TableBundleSnapshot;

      const extrasFileName = 'data/' + nodeId + '.extras.json';
      let extras;
      if (unzipped[extrasFileName] !== undefined) {
        // @ts-ignore
        extras = fflate.strFromU8(unzipped[extrasFileName]);
      }
      const dataSheet = new TableBundleDataSheet(sn, extras);
      this._nodeMap.set(nodeId, dataSheet);
    }
  }

}

export interface ITableBundleInitOptions {
  loader: ITableBundleLoader;
  saver: ITableBundleSaver;
}

export interface ITableBundleLoader{
    load(path: string): Buffer;
}
export interface ITableBundleSaver{
    save(bundle: Uint8Array, path: string): void;
}
