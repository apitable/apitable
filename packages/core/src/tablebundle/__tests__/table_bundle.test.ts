import { ITableBundleInitOptions, ITableBundleLoader, ITableBundleSaver, TableBundle } from '../table_bundle';
import { TableBundleSnapshot } from '../table_bundle_snapshot';
import { TableBundleDataSheet } from '../table_bundle_datasheet';
import * as fs from 'fs';

const assetFilePath = `${__dirname}/test.zip`;

const transformSnapshot = (data: any): TableBundleSnapshot => {
  if (typeof data === 'string') {
    return JSON.parse(data) as TableBundleSnapshot;
  }
  return data as TableBundleSnapshot;
};

class TestLoader implements ITableBundleLoader {
  load(path: string): Buffer {
    return fs.readFileSync(path);
  }
}
class TestSaver implements ITableBundleSaver {
  save(array: Uint8Array, path: string): void {
    return fs.writeFileSync(path, array);
  }

}
const generateOption = (): ITableBundleInitOptions => {
  const options: ITableBundleInitOptions = {
    loader: new TestLoader(),
    saver: new TestSaver()
  };
  return options;
};

describe('TableBundle', () => {
  it('add data sheet', () => {
    const tableBundle = TableBundle.new(generateOption());
    tableBundle.apply(new TableBundleDataSheet(transformSnapshot(snapshot)));
    expect(tableBundle.nodeMap().size).toEqual(1);
    expect(tableBundle.nodeMap()).toMatchSnapshot();
    expect(tableBundle.getDataSheet('dst001')).toMatchSnapshot();
  });

  it('generate zip file', () => {
    const tableBundle = TableBundle.new(generateOption());
    tableBundle.apply(new TableBundleDataSheet(transformSnapshot(snapshot)));
    tableBundle.save(assetFilePath);

    expect(fs.existsSync(assetFilePath)).toBe(true);
  });

  it('parser zip file', () => {
    expect(fs.existsSync(assetFilePath)).toBe(true);
    const tableBundle = TableBundle.new(generateOption());
    tableBundle.loadFile(assetFilePath);
    expect(tableBundle.nodeMap().size).toEqual(1);
    expect(tableBundle.nodeMap()).toMatchSnapshot();
    expect(tableBundle.getDataSheet('dst001')).toMatchSnapshot();
    fs.unlinkSync(assetFilePath);
  });

});

const snapshot = `
{
\t"recordMap": {
\t\t"fldMNKMZ7X4P6": {
\t\t\t"data": {
\t\t\t\t"flds5VCUKUZMH": ["optQAZBO9LO3t"],
\t\t\t\t"fldMNKMZ7X4P6": [{
\t\t\t\t\t"type": 1,
\t\t\t\t\t"text": "x z c x z c "
\t\t\t\t}],
\t\t\t\t"fldt3qT5kfsBk": [{
\t\t\t\t\t"mimeType": "image/png",
\t\t\t\t\t"token": "space/2023/05/17/0026b58ee1654c439bffc3974f9cb4dc",
\t\t\t\t\t"bucket": "QNY1",
\t\t\t\t\t"size": 60983,
\t\t\t\t\t"name": "image.png",
\t\t\t\t\t"width": 856,
\t\t\t\t\t"id": "atcW6tzBymOzF",
\t\t\t\t\t"height": 646
\t\t\t\t}]
\t\t\t},
\t\t\t"id": "recXXXX",
\t\t\t"recordMeta": null
\t\t}
\t},
\t"meta": {
\t\t"fieldMap": {
\t\t\t"flds5VCUKUZMH": {
\t\t\t\t"type": 4,
\t\t\t\t"name": "选项",
\t\t\t\t"property": {
\t\t\t\t\t"options": [{
\t\t\t\t\t\t"color": 0,
\t\t\t\t\t\t"name": "想啊啥",
\t\t\t\t\t\t"id": "optQAZBO9LO3t"
\t\t\t\t\t}]
\t\t\t\t},
\t\t\t\t"id": "flds5VCUKUZMH"
\t\t\t},
\t\t\t"fldMNKMZ7X4P6": {
\t\t\t\t"type": 19,
\t\t\t\t"name": "标题",
\t\t\t\t"property": {
\t\t\t\t\t"defaultValue": ""
\t\t\t\t},
\t\t\t\t"id": "fldMNKMZ7X4P6"
\t\t\t},
\t\t\t"fldt3qT5kfsBk": {
\t\t\t\t"type": 6,
\t\t\t\t"name": "附件",
\t\t\t\t"id": "fldt3qT5kfsBk"
\t\t\t}
\t\t},
\t\t"views": [{
\t\t\t"frozenColumnCount": 1,
\t\t\t"columns": [{
\t\t\t\t"statType": 1,
\t\t\t\t"fieldId": "fldMNKMZ7X4P6"
\t\t\t}, {
\t\t\t\t"fieldId": "flds5VCUKUZMH"
\t\t\t}, {
\t\t\t\t"fieldId": "fldt3qT5kfsBk"
\t\t\t}],
\t\t\t"rows": [{
\t\t\t\t"recordId": "recfte8MK9TWd"
\t\t\t}, {
\t\t\t\t"recordId": "recc2klate85d"
\t\t\t}, {
\t\t\t\t"recordId": "recMWYWw95sVi"
\t\t\t}],
\t\t\t"type": 1,
\t\t\t"autoSave": false,
\t\t\t"name": "维格视图",
\t\t\t"id": "viwruy3vMmo70"
\t\t}]
\t}

}
`;