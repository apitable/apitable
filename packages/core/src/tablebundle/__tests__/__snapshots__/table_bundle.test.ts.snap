// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TableBundle add data sheet 1`] = `
Map {
  "dst001" => TableBundleDataSheet {
    "extras": undefined,
    "snapshot": {
      "meta": {
        "fieldMap": {
          "fldMNKMZ7X4P6": {
            "id": "fldMNKMZ7X4P6",
            "name": "标题",
            "property": {
              "defaultValue": "",
            },
            "type": 19,
          },
          "flds5VCUKUZMH": {
            "id": "flds5VCUKUZMH",
            "name": "选项",
            "property": {
              "options": [
                {
                  "color": 0,
                  "id": "optQAZBO9LO3t",
                  "name": "想啊啥",
                },
              ],
            },
            "type": 4,
          },
          "fldt3qT5kfsBk": {
            "id": "fldt3qT5kfsBk",
            "name": "附件",
            "type": 6,
          },
        },
        "views": [
          {
            "autoSave": false,
            "columns": [
              {
                "fieldId": "fldMNKMZ7X4P6",
                "statType": 1,
              },
              {
                "fieldId": "flds5VCUKUZMH",
              },
              {
                "fieldId": "fldt3qT5kfsBk",
              },
            ],
            "frozenColumnCount": 1,
            "id": "viwruy3vMmo70",
            "name": "维格视图",
            "rows": [
              {
                "recordId": "recfte8MK9TWd",
              },
              {
                "recordId": "recc2klate85d",
              },
              {
                "recordId": "recMWYWw95sVi",
              },
            ],
            "type": 1,
          },
        ],
      },
      "recordMap": {
        "fldMNKMZ7X4P6": {
          "data": {
            "fldMNKMZ7X4P6": [
              {
                "text": "x z c x z c ",
                "type": 1,
              },
            ],
            "flds5VCUKUZMH": [
              "optQAZBO9LO3t",
            ],
            "fldt3qT5kfsBk": [
              {
                "bucket": "QNY1",
                "height": 646,
                "id": "atcW6tzBymOzF",
                "mimeType": "image/png",
                "name": "image.png",
                "size": 60983,
                "token": "space/2023/05/17/0026b58ee1654c439bffc3974f9cb4dc",
                "width": 856,
              },
            ],
          },
          "id": "recXXXX",
          "recordMeta": null,
        },
      },
    },
  },
}
`;

exports[`TableBundle add data sheet 2`] = `
TableBundleDataSheet {
  "extras": undefined,
  "snapshot": {
    "meta": {
      "fieldMap": {
        "fldMNKMZ7X4P6": {
          "id": "fldMNKMZ7X4P6",
          "name": "标题",
          "property": {
            "defaultValue": "",
          },
          "type": 19,
        },
        "flds5VCUKUZMH": {
          "id": "flds5VCUKUZMH",
          "name": "选项",
          "property": {
            "options": [
              {
                "color": 0,
                "id": "optQAZBO9LO3t",
                "name": "想啊啥",
              },
            ],
          },
          "type": 4,
        },
        "fldt3qT5kfsBk": {
          "id": "fldt3qT5kfsBk",
          "name": "附件",
          "type": 6,
        },
      },
      "views": [
        {
          "autoSave": false,
          "columns": [
            {
              "fieldId": "fldMNKMZ7X4P6",
              "statType": 1,
            },
            {
              "fieldId": "flds5VCUKUZMH",
            },
            {
              "fieldId": "fldt3qT5kfsBk",
            },
          ],
          "frozenColumnCount": 1,
          "id": "viwruy3vMmo70",
          "name": "维格视图",
          "rows": [
            {
              "recordId": "recfte8MK9TWd",
            },
            {
              "recordId": "recc2klate85d",
            },
            {
              "recordId": "recMWYWw95sVi",
            },
          ],
          "type": 1,
        },
      ],
    },
    "recordMap": {
      "fldMNKMZ7X4P6": {
        "data": {
          "fldMNKMZ7X4P6": [
            {
              "text": "x z c x z c ",
              "type": 1,
            },
          ],
          "flds5VCUKUZMH": [
            "optQAZBO9LO3t",
          ],
          "fldt3qT5kfsBk": [
            {
              "bucket": "QNY1",
              "height": 646,
              "id": "atcW6tzBymOzF",
              "mimeType": "image/png",
              "name": "image.png",
              "size": 60983,
              "token": "space/2023/05/17/0026b58ee1654c439bffc3974f9cb4dc",
              "width": 856,
            },
          ],
        },
        "id": "recXXXX",
        "recordMeta": null,
      },
    },
  },
}
`;

exports[`TableBundle parser zip file 1`] = `
Map {
  "dst001" => TableBundleDataSheet {
    "extras": undefined,
    "snapshot": {
      "meta": {
        "fieldMap": {
          "fldMNKMZ7X4P6": {
            "id": "fldMNKMZ7X4P6",
            "name": "标题",
            "property": {
              "defaultValue": "",
            },
            "type": 19,
          },
          "flds5VCUKUZMH": {
            "id": "flds5VCUKUZMH",
            "name": "选项",
            "property": {
              "options": [
                {
                  "color": 0,
                  "id": "optQAZBO9LO3t",
                  "name": "想啊啥",
                },
              ],
            },
            "type": 4,
          },
          "fldt3qT5kfsBk": {
            "id": "fldt3qT5kfsBk",
            "name": "附件",
            "type": 6,
          },
        },
        "views": [
          {
            "autoSave": false,
            "columns": [
              {
                "fieldId": "fldMNKMZ7X4P6",
                "statType": 1,
              },
              {
                "fieldId": "flds5VCUKUZMH",
              },
              {
                "fieldId": "fldt3qT5kfsBk",
              },
            ],
            "frozenColumnCount": 1,
            "id": "viwruy3vMmo70",
            "name": "维格视图",
            "rows": [
              {
                "recordId": "recfte8MK9TWd",
              },
              {
                "recordId": "recc2klate85d",
              },
              {
                "recordId": "recMWYWw95sVi",
              },
            ],
            "type": 1,
          },
        ],
      },
      "recordMap": {
        "fldMNKMZ7X4P6": {
          "data": {
            "fldMNKMZ7X4P6": [
              {
                "text": "x z c x z c ",
                "type": 1,
              },
            ],
            "flds5VCUKUZMH": [
              "optQAZBO9LO3t",
            ],
            "fldt3qT5kfsBk": [
              {
                "bucket": "QNY1",
                "height": 646,
                "id": "atcW6tzBymOzF",
                "mimeType": "image/png",
                "name": "image.png",
                "size": 60983,
                "token": "space/2023/05/17/0026b58ee1654c439bffc3974f9cb4dc",
                "width": 856,
              },
            ],
          },
          "id": "recXXXX",
          "recordMeta": null,
        },
      },
    },
  },
}
`;

exports[`TableBundle parser zip file 2`] = `
TableBundleDataSheet {
  "extras": undefined,
  "snapshot": {
    "meta": {
      "fieldMap": {
        "fldMNKMZ7X4P6": {
          "id": "fldMNKMZ7X4P6",
          "name": "标题",
          "property": {
            "defaultValue": "",
          },
          "type": 19,
        },
        "flds5VCUKUZMH": {
          "id": "flds5VCUKUZMH",
          "name": "选项",
          "property": {
            "options": [
              {
                "color": 0,
                "id": "optQAZBO9LO3t",
                "name": "想啊啥",
              },
            ],
          },
          "type": 4,
        },
        "fldt3qT5kfsBk": {
          "id": "fldt3qT5kfsBk",
          "name": "附件",
          "type": 6,
        },
      },
      "views": [
        {
          "autoSave": false,
          "columns": [
            {
              "fieldId": "fldMNKMZ7X4P6",
              "statType": 1,
            },
            {
              "fieldId": "flds5VCUKUZMH",
            },
            {
              "fieldId": "fldt3qT5kfsBk",
            },
          ],
          "frozenColumnCount": 1,
          "id": "viwruy3vMmo70",
          "name": "维格视图",
          "rows": [
            {
              "recordId": "recfte8MK9TWd",
            },
            {
              "recordId": "recc2klate85d",
            },
            {
              "recordId": "recMWYWw95sVi",
            },
          ],
          "type": 1,
        },
      ],
    },
    "recordMap": {
      "fldMNKMZ7X4P6": {
        "data": {
          "fldMNKMZ7X4P6": [
            {
              "text": "x z c x z c ",
              "type": 1,
            },
          ],
          "flds5VCUKUZMH": [
            "optQAZBO9LO3t",
          ],
          "fldt3qT5kfsBk": [
            {
              "bucket": "QNY1",
              "height": 646,
              "id": "atcW6tzBymOzF",
              "mimeType": "image/png",
              "name": "image.png",
              "size": 60983,
              "token": "space/2023/05/17/0026b58ee1654c439bffc3974f9cb4dc",
              "width": 856,
            },
          ],
        },
        "id": "recXXXX",
        "recordMeta": null,
      },
    },
  },
}
`;
